### PR TITLE
Add interpreter for BroadcastInDimOp

### LINKED ISSUE #964: Add interpreter for broadcast_in_dim
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) operand tensor.
(I2) broadcast_dimensions 1-dimensional tensor constant of type `si64`.
(C1) `operand` and `result` have the same element type.
(C2) size(`broadcast_dimensions`) $=$ rank(`operand`).
(C3) $0 \le$ `broadcast_dimensions[i]` $\lt$ rank(`result`) for all
     dimensions i in `operand`.
(C4) All dimensions in `broadcast_dimensions` are unique.
(C5) For all dimensions `j` in `operand`:
     * `dim(operand, j) = 1` or
     * `dim(operand, j) = dim(result, i)` where `i = broadcast_dimensions[j]`.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) operand is not a tensor. (Covered by ODS).
I2: a) broadcast_dimensions is not a 1-dimensional tensor.
    b) broadcast_dimensions does not have `si64` type. (Covered by ODS).
C1: a) element_type(operand) != element_type(result). (Covered by ODS).
C2: a) size(broadcast_dimensions) != rank(operand).
C3: a) broadcast_dimensions[i] < 0 for any i.
    b) broadcast_dimensions[i] >= rank(result) for any i.
C4: a) `broadcast_dimensions` values are not unique.
C5: a) dim(operand, j) != 1 and dim(operand, j) != dim(result, i) where `i = broadcast_dimensions[j]` for all dimensions `j` in `operand`.
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
I2a: broadcast_dimensions is not a 1-dimensional tensor.
C2a: size(broadcast_dimensions) != rank(operand).
C3a: broadcast_dimensions[i] < 0 for any i.
C3b: broadcast_dimensions[i] >= rank(result) for any i.
C4a: `broadcast_dimensions` values are not unique.
C5a: dim(operand, j) != 1 and dim(operand, j) != dim(result, i) where `i = broadcast_dimensions[j]` for all dimensions `j` in `operand`.
```

closes #964

### CODE REVIEW

File: stablehlo/tests/interpret_broadcast_in_dim.mlir [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  // CHECK-NEXT: 1 : i64
+  // CHECK-NEXT: 2 : i64
+  // CHECK-NEXT: 2 : i64
+  // CHECK-NEXT: 2 : i64
+  // CHECK-NEXT: 2 : i64
+  // CHECK-NEXT: 3 : i64
+  // CHECK-NEXT: 3 : i64
+  // CHECK-NEXT: 3 : i64
+  // CHECK-NEXT: 3 : i64
+}
```
Comment (zhouxin913) [inline stablehlo/tests/interpret_broadcast_in_dim.mlir:23]: Could you also add a comment to print the matrix in easy to understand form just for readability ?

**Review by zhouxin913 — APPROVED** [review]

File: stablehlo/tests/interpret_broadcast_in_dim.mlir (hunk 1 above)
Comment (ghpvnist) [reply stablehlo/tests/interpret_broadcast_in_dim.mlir:23]: Once #1026 is merged, these files will be replaced with the new testing structure, and here, we should be able to format the result in the `almost_eq` op.

**Review by zhouxin913 — APPROVED** [review]

File: stablehlo/dialect/StablehloOps.td [hunk 2]
```diff
@@ -1844,8 +1844,8 @@ def StableHLO_BroadcastOp : StableHLO_ShapedInterfaceOp<"broadcast",
 }
 
 def StableHLO_BroadcastInDimOp : StableHLO_Op<"broadcast_in_dim",
-      [Pure, SameOperandsAndResultElementType]> {
-  let summary = "Broadcast a tensor into the given shape by adding dimensions.";
+      [Pure, SameOperandsAndResultElementType /*broadcast_in_dim_c1*/]> {
+  let summary = "BroadcastInDim operation";
```
Comment (sdasgup3) [inline stablehlo/dialect/StablehloOps.td:1848]: Add input constraints as per  the checklist.

File: stablehlo/tests/ops_stablehlo.mlir [hunk 3]
```diff
@@ -873,54 +857,70 @@ func.func @dynamic_broadcast_in_dim_shape_mismatch(%arg0: tensor<32xf32>, %shape
...
   %0 = "stablehlo.broadcast_in_dim"(%arg0) {broadcast_dimensions = dense<[0,1,2]> : tensor<3xi64>} : (tensor<1x2x3xi32>) -> tensor<3xi32>
   func.return %0 : tensor<3xi32>
 }
 
 // -----
 
-func.func @broadcast_in_dim_duplicate_bcast_dimensions(%arg0: tensor<1x1x3xi32>) -> tensor<1x2x3xi32> {
-  // expected-error@+1 {{broadcast_dimensions should not have duplicates}}
-  %0 = "stablehlo.broadcast_in_dim"(%arg0) {broadcast_dimensions = dense<[0,0,2]> : tensor<3xi64>} : (tensor<1x1x3xi32>) -> tensor<1x2x3xi32>
+func.func @broadcast_in_dim_c3(%arg0: tensor<1x2xi32>) -> tensor<1x2x3xi32> {
```
Comment (burmako) [inline stablehlo/tests/ops_stablehlo.mlir:892]: Both this test and the previous test seem to cover C3b. Let's drop one of them?

File: stablehlo/tests/ops_stablehlo.mlir [hunk 4]
```diff
@@ -945,6 +945,14 @@ func.func @broadcast_in_dim_unranked_operand(%arg0 : tensor<*xf32>) -> tensor<2x
 
 // -----
 
+// CHECK-LABEL: func @broadcast_in_dim_zero_rank
```
Comment (burmako) [inline stablehlo/tests/ops_stablehlo.mlir:948]: Do we need this as a separate positive test?

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -54,6 +54,20 @@ Tensor evalAndOp(const Tensor &lhs, const Tensor &rhs, Type resultType) {
...
+Tensor evalBroadcastInDimOp(const Tensor &operand,
+                            ArrayRef<int64_t> broadcastDimensions,
+                            Type resultType) {
+  Tensor result(resultType);
+  auto operandShape = operand.getType().getShape();
+  for (auto it = result.index_begin(); it != result.index_end(); ++it) {
+    SmallVector<int64_t> inIndices;
+    for (auto [inDim, outDim] : llvm::enumerate(broadcastDimensions))
+      inIndices.push_back(operandShape[inDim] == 1 ? 0 : (*it)[outDim]);
+    result.set(*it, operand.get(inIndices));
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:66]: Feedback on names:
  1) `it` => `resultIt` for consistency with other eval functions.
  2) `inIndices` => `operandIdx` for consistency with other eval functions.
  3) `inDim` => `operandDim` to follow ODS names.
  4) `outDim` => `resultDim` to follow ODS names.

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/tests/ops_stablehlo.mlir (hunk 4 above)
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:948]: No, not necessary, I'll remove this.

**Review by burmako — APPROVED** [review]

**Pull request merged**